### ISSUE TITLE
Some risks regarding dll entry point scope

### ISSUE THREAD

**Issue by greenozon** [Luca1991/DiscCheckEmu#20]
This is old topic, it might hit sooner or later...
so it's all about the limitations which are exposed inside the dll entry point context.

According to very well written [article ](https://learn.microsoft.com/en-us/windows/win32/dlls/dynamic-link-library-best-practices?redirectedfrom=MSDN)
it's risky to call routines from user32.dll  including MessageBox one

The recommendation would be to replace it with OutputDebugString or so..

**Comment by Luca1991** [comment]
Hi @greenozon 

Replying from my mobile, so please forgive me for any inaccuracies.

Aren't we using MessageBox to notify the user about a critical (unrecoverable) error? (At least in release builds). Even so, yes I agree that it is not the best practice as you pointed out. What is a valid alternative? OutputDebugString requires a debugger attached, something that of curses isn't the case here.

**Comment by greenozon** [comment]
https://stackoverflow.com/questions/8377569/howto-call-messagebox-in-dllmain

In ideal case dll EP should be as small as possible (according to OS limitations put over it)
so I'm thinking about adding export API that one could call directly..
in this case you are in a clean position

**Comment by Luca1991** [comment]
> so I'm thinking about adding export API that one could call directly..

Called from where?

**Comment by greenozon** [comment]
I mean you are doing a lot of business logic (lets call it so) inside the 
```
BOOL WINAPI DllMain()
```

right now it works, but according to official MS docs this scope is super limited and general recommendation is to leave it empty

thus, the proposal is to add export API (name it init, whatever) and then call it later on after you load the dll itself

**Comment by Luca1991** [comment]
> right now it works, but according to official MS docs this scope is super limited and general recommendation is to leave it empty

yeah, I understand what do you mean, but this is how MS uses Detorus in their examples. I don't know if adding an export is a good idea here honestly.

**Comment by greenozon** [comment]
Yeah, I've explored the code and you are right
the only moment you have is just one point in  time - 
CreateRemoteThread()
you are loading your dll which we are discussing above and by using LoadLibrary you automatically get called by the DLL entry point, so you don't have 2nd chance to call your specific exported API...

I was thinking smth like spin off a new thread that will run later on...
here is an example:  https://devblogs.microsoft.com/oldnewthing/20070904-00/?p=25283

another interesting approach found over here... https://www.codeproject.com/Articles/20084/A-More-Complete-DLL-Injection-Solution-Using-Creat

**Comment by Luca1991** [comment]
Sorry for the late reply.
The problem with a new thread is that we need do be 100% sure that it is synchronized correctly before resuming the process execution (from the injector).
I had purposely used a (bad) Sleep from the injector to let the dll finish installing its hooks and my plan is to change it with something cleaner. If we are going to create another thread from the dll itself (also discouraged by Raymond Chen), we need to add another way to make sure the hooks are installed before resuming the main process execution. 
In the second link, the idea to use an export seems better... I wonder if there might be some compatibility issues with this, tho.

**Comment by greenozon** [comment]
Agree, spinning off a thread from dll entry point is also not a great idea...

**Comment by Luca1991** [comment]
Hi @greenozon 

In v0.3.0 I finally removed the (bad) Sleep from DCELoader, replacing it with a (propper) WaitForSingleObject call.
Regarding the export: after some reasearch and testing I don't think that it is actually needed! 
When  the DCEAPIHook dll is loaded, the process is actually in suspended state and it will be resumed only after the WaitForSingleObject or a 10s timeout (failure). In this scenario I don't think that any deadlock could happen.
In DCE v0.4.0 I added in-memory patches support: I tried applying many patches without any issue (nor deadlocks).

I'm closing this issue. If problems arise in the future (I'm confident that this won't be the case given that the process is suspended anyway), I will reopen it and we'll see what to do ;)
